provisioning/miniprov/hare_mp/main.py

```diff
@@ -254,21 +255,202 @@ def test(args):
         exit(-1)
 
 
+def executecmds(cmd: List[str]) -> list:
+    process = subprocess.Popen(cmd,
+                               stdin=subprocess.PIPE,
+                               stdout=subprocess.PIPE,
+                               stderr=subprocess.PIPE,
+                               encoding='utf8')
+    resp = list(process.communicate())
+    resp.append(str(process.returncode))
+
+    return resp
+
+
+def test_hare_prereq():
+    """Test suite for setting the test environment before start of test."""
+    pcs_status = ['pcs', 'status']
+    cluster_stop = ['cortx', 'cluster', 'stop']
+
+    logging.info('Test started on Host: {}'.format(executecmds(['hostname'])))
+    logging.info('Check that all services are up in PCS.')
+
+    resp = executecmds(pcs_status)
+    logging.info('PCS status: %s', resp[0])
+
+    if 'cluster is not currently' in resp[0]:
+        return
+
+    logging.info('Make Node ready for testing, by stopping the cluster')
+    resp = executecmds(cluster_stop)
+    if int(resp[2]):
+        logging.info('Cluster status : %s', resp[0])
+        logging.error('Cluster failed to stop %s', resp[1])
+        raise Exception(
+            f'Command {cluster_stop} exited with error code '
+            f'{int(resp[2])}. Command output: {resp[1]}')
+    else:
+        logging.info('Cluster Stopped : %s', resp[0])
+        for line in resp[0]:
+            assert 'Cluster stop is in progress' not in line, \
+                'Cluster is in progress.' if 'Cluster is in progress.' \
+                else line
+
+
+def test_hare_postreq(cdf_file: str, timeinfo: str, logfile: str):
+    """Test suite is for restoring the setting after test."""
+    pcs_status = ['pcs', 'status']
+    cluster_start = ['cortx', 'cluster', 'start']
+    hctl_status = ['hctl', 'status', '-d']
+
+    logging.info('Start the Cluster')
+    resp = executecmds(cluster_start)
+    logging.info('cluster status: %s', resp[0])
+    if int(resp[2]):
+        logging.error('Cluster failed to start %s', resp[1])
+        resp = executecmds(['journalctl', '--since', timeinfo, '>', logfile])
+        logging.info('created journal log %s', logfile)
+        raise Exception(
+            f'Command {cluster_start} exited with error code {int(resp[2])}'
+            f'Command output: {resp[1]}')
+    else:
+        for line in resp[0]:
+            assert 'Cluster start operation performed' not in line, \
+                'Cluster not yet started.' if 'Cluster not yet started.' \
+                else line
+
+    cluster_sts = check_cluster_status(cdf_file)
+    if cluster_sts:
+        logging.error('Cluster status reports failure')
+        resp = executecmds(['journalctl', '--since', timeinfo, '>', logfile])
+        logging.info('created journal log.%s', logfile)
+        exit(-1)
+
+    logging.info('PCS: Check all services are up.')
+    sleep(20)
```

Comment:
Sleep-driven development is error-prone.

FYI: in the past I implemented a tool called `pcswrap` that can be helpful for things like this (e.g. start all services and wait until they are started for sure - with timeout). Currently its source code is in cortx-ha repository: https://github.com/Seagate/cortx-ha/tree/main/ha/pcswrap

We could leverage either the whole pcswrap tool (TBD) or reuse some parts of its codebase. For instance, [Waiter](https://github.com/Seagate/cortx-ha/blob/19c6cad22b4a7f2036bc7e292b40383bae4d7fdb/ha/pcswrap/pcswrap/internal/waiter.py#L21) class can be beneficial:

https://github.com/Seagate/cortx-ha/blob/19c6cad22b4a7f2036bc7e292b40383bae4d7fdb/ha/pcswrap/pcswrap/client.py#L97-L103


@mssawant What's your opinion? Can we take back (or copy) pcswrap back to our repository?